security/pkg/workload/server.go

```diff
@@ -46,6 +48,9 @@ type SDSServer struct {
 
 	// The grpc server that listens on the above udsPaths.
 	grpcServer *grpc.Server
+
+	// secret service that serves certificate and private key
+	secretServer SecretServer
```

Comment:
My initial design is actually having SDSServer implement SecretServer.
Node agent instantiates a SDSServer (as a subclass of SecretServer) for each workload, and puts the workload key/cert into the SDSServer through SetServiceIdentityPrivateKey() and SetServiceIdentityCert().
After receiving the workload key/cert, the SDSServer runs as a goroutine that exposes the SDS API to the workload.
When I first wrote this SecretServer code I didn't know about the details of SDS. But it seems the original design can still apply here. Does this make sense?

Reply:
To clarify, the SDSServer is designed to be sharable for different workloads. SDSServer can listen at different UDS paths

Reply:
after looking at the SecretServer interface: https://github.com/mangchiandjjoe/istio/blob/557a670a8ab3ba920d483b3126bcc3232e96e242/security/pkg/workload/secretserver.go#L22

I am not very convinced that we should fit SDS there. My understanding is that SDS already provides the sufficient interface between nodeagent and workload. And we may deprecate the file based solution for distributing key/cert in the future. Private key should only be stored in memory (maybe in vTPM in the future)

Reply:
It does make sense. I update the SDSServer object.

Reply:
@wenchenglu 
I think the SecreteServer interface is for node agent to set key/cert into the SecretServer object (this is happening inside node agent binary). The SDS interface is to expose SDS API to Envoy (this is happening over the UDS).

If one SDSServer serves multiple workloads, can we modify the interface a bit for this? Such as SetServiceIdentityPrivateKey(key, path)?
I'm a little keen to using interfaces. Even though we are going to get rid of files, it could support mocking in our tests :)

Reply:
that makes sense. You may want to update the description for SecretServer interface which gives me the impression this is from handling the communication between node agent and workload, which is the responsibility for SDS API.